src/kernel/thread.c

```diff
@@ -372,14 +380,14 @@ void schedule(void)
                 SCHED_ENQUEUE(candidate);
                 /* we can't, need to reschedule */
                 NODE_STATE(ksSchedulerAction) = SchedulerAction_ChooseNewThread;
-                scheduleChooseNewThread();
+                switchedDomain = scheduleChooseNewThread();
             } else if (was_runnable && candidate->tcbPriority == NODE_STATE(ksCurThread)->tcbPriority) {
                 /* We append the candidate at the end of the scheduling queue, that way the
                  * current thread, that was enqueued at the start of the scheduling queue
                  * will get picked during chooseNewThread */
                 SCHED_APPEND(candidate);
                 NODE_STATE(ksSchedulerAction) = SchedulerAction_ChooseNewThread;
-                scheduleChooseNewThread();
+                switchedDomain = scheduleChooseNewThread();
             } else {
                 assert(candidate != NODE_STATE(ksCurThread));
                 switchToThread(candidate);
```

Comment:
It appears that this call to `switchToThread` will occur without checking that the current domain has not expired. If it is the case that the domain has expired, then this thread will be scheduled and the later call to `commitTime` will send `domainTime` to zero before the kernel exits. If I've understood this correctly, then two possible solutions are:
1. `possibleSwitchTo` can check `isCurDomainExpired` so that `ksSchedulerAction` can never point to a thread if the domain has expired.
2. schedule can check `isCurDomainExpired` and sensibly avoid this call to `switchToThread`
Thoughts?

Reply:
There is a `rescheduleRequired` in checkBudget that is called in checkBudget.  Is it possible for this to not have been called by the time schedule is called such that scheduleChooseNewThread isn't called?

Reply:
Great point, will check it out.

Reply:
Your advice was excellent and appears to account for the majority of cases. There appears to be one path which is a bit uncertain: when the kernel is preempted and `mcsIRQ` is called. Is it clear what should happen if the current thread is not "schedulable" and the current scheduling context is not "valid"? In that case neither `checkBudget` nor `chargeBudget` are called and I can see the following:
* the scheduler action could not be `ResumeCurrentThread` GOOD
* if the scheduler action is `ChooseNewThread` then schedule will call `scheduleChooseNewThread` GOOD
* if the schedule action contains a candidate, then 2/3 cases will call `scheduleChooseNewThread` GOOD, but what about the third case when the kernel makes the switch to the proposed candidate?

Reply:
I think it should be covered by the checkBudget located in preemptionPoint().  If EXCEPTION_PREEMPTED is returned (which should be the only condition that causes mcsIRQ to be called), then it is either because an IRQ is pending or because checkBudget was called and returned false.  I'm not sure what preemptionPoint is supposed to ensure here though.  I believe it should be equivalent to what would happen upon entry to the kernel and so should probably be updating the timestamp and checking the budget unconditionally if `ksWorkUnitsCompleted >= CONFIG_MAX_NUM_WORK_UNITS_PER_PREEMPTION` is true